src/java/org/apache/cassandra/index/sai/disk/v1/PrimaryKeyMap.java

```diff
@@ -54,42 +54,47 @@ default void close() throws IOException
 
     class DefaultPrimaryKeyMap implements PrimaryKeyMap
     {
-        private final long size;
         private final FileHandle primaryKeysFile;
         private final FileHandle primaryKeyOffsetsFile;
         private final RandomAccessReader primaryKeys;
         private final LongArray primaryKeyOffsets;
+        private final IndexComponents indexComponents;
+        private final NumericValuesMeta metadata;
 
         private final PrimaryKey.PrimaryKeyFactory keyFactory;
 
         public DefaultPrimaryKeyMap(IndexComponents indexComponents, TableMetadata tableMetadata) throws IOException
         {
+            this.indexComponents = indexComponents;
+
             MetadataSource metadataSource = MetadataSource.loadGroupMetadata(indexComponents);
-            NumericValuesMeta meta = new NumericValuesMeta(metadataSource.get(IndexComponents.PRIMARY_KEY_OFFSETS.name));
-            this.size = meta.valueCount;
+            this.metadata = new NumericValuesMeta(metadataSource.get(IndexComponents.PRIMARY_KEY_OFFSETS.name));
 
             this.primaryKeyOffsetsFile = indexComponents.createFileHandle(IndexComponents.PRIMARY_KEY_OFFSETS);
-            this.primaryKeyOffsets = new MonotonicBlockPackedReader(primaryKeyOffsetsFile, indexComponents, meta).open();
+            this.primaryKeyOffsets = new MonotonicBlockPackedReader(primaryKeyOffsetsFile, indexComponents, metadata)
+                    .open();
 
             this.primaryKeysFile = indexComponents.createFileHandle(IndexComponents.PRIMARY_KEYS);
             this.primaryKeys = primaryKeysFile.createReader();
 
             this.keyFactory = PrimaryKey.factory(tableMetadata);
         }
 
-        private DefaultPrimaryKeyMap(DefaultPrimaryKeyMap orig)
+        private DefaultPrimaryKeyMap(DefaultPrimaryKeyMap orig) throws IOException
         {
+            this.indexComponents = orig.indexComponents;
+            this.metadata = orig.metadata;
+
             this.primaryKeysFile = orig.primaryKeysFile.sharedCopy();
             this.primaryKeyOffsetsFile = orig.primaryKeyOffsetsFile.sharedCopy();
 
             this.primaryKeys = primaryKeysFile.createReader();
-            this.primaryKeyOffsets = orig.primaryKeyOffsets;
+            this.primaryKeyOffsets = new MonotonicBlockPackedReader(primaryKeyOffsetsFile, indexComponents, metadata)
```

Comment:
I'm also interested to know why we can't use the original reader in the copy.

Reply:
The problem I can see is that the `MonotonicBlockPackedReader` stores the `FileHandle` as a field. 
Unfortunately each (copied) instance of `PrimaryKeyMap` gets its own `FileHandle` copies. 
If we just passed the MBPR instance directly to each copy of the `PrimaryKeyMap`, the MBPR's `FileHandle` would not match the one in the PKM. Closing the original (first) PrimaryKeyMap before one of its copies would close the `FileHandle` used by the MBPR in the copies and I guess it could pose a problem.